### DISCUSSION TITLE
Need help with EthereumAbiFunction

### DISCUSSION THREAD

**Discussion by MhAhmadAli** [trustwallet/wallet-core#3355] [Q&A]
I am trying to write a code for follwing function:

`exactInputSingle((address,address,uint24,address,uint256,uint256,uint160))`

As you can see, the function needs a tuple as input.

I have tried using "addParamArray()" to add an array but the function signature doesn't match when I look at the encoded output.

Required Function Signature: 0x04e45aaf

**Comment by satoshiotomakan** [comment]
Hi @MhAhmadAli, could you please share the code?

**Comment by MhAhmadAli** [comment]
This is what I am currently on right now @satoshiotomakan 

```
val swapAddress = "0x365C5F0B816828936320ea143e337fbA7D1b911E"
val tokenIn = "0x559bb8EEa09601d9AC65d799808142aEeA60e37E"
val tokenOut = "0x30ec47F7DFae72eA79646e6cf64a8A7db538915b"
val fee = 500
val amountIn = 1000000000000000000
val amountOutMin = 0
val sqrtPriceLimitX96 = 0

val exactInput = EthereumAbiFunction("exactInputSingle")
exactInput.addParamArray(false)
exactInput.addInArrayParamAddress(0, tokenIn.toByteArray())
exactInput.addInArrayParamAddress(1, tokenOut.toByteArray())
exactInput.addInArrayParamUIntN(2, 24, BigInteger.valueOf(fee.toLong()).toByteArray())
exactInput.addInArrayParamAddress(3, receiverAddress.toByteArray())
exactInput.addInArrayParamUInt256(4, BigInteger.valueOf(amountIn).toByteArray())
exactInput.addInArrayParamUInt256(5, BigInteger.valueOf(amountOutMin.toLong()).toByteArray())
exactInput.addInArrayParamUIntN(6, 160, BigInteger.valueOf(sqrtPriceLimitX96.toLong()).toByteArray())

println("exactInput: ${EthereumAbi.encode(exactInput).toHexString()}")
```

Function for converting encoded data to Hex string:
```
private fun ByteArray.toHexString(withPrefix: Boolean = true): String {
        val stringBuilder = StringBuilder()
        if (withPrefix) {
            stringBuilder.append("0x")
        }
        for (element in this) {
            stringBuilder.append(String.format("%02x", element and 0xFF.toByte()))
        }
        return stringBuilder.toString()
    }
```


I have tried it the normal way without using an array as well.

**Reply by satoshiotomakan** [reply]
Thank you, as I can see, we don't provide a function to specify a tuple at this moment.
We plan to rework the existing Ethereum ABI API, there should be the tuple type parameter.

**Reply by MhAhmadAli** [reply]
Any ETA on that?

**Reply by satoshiotomakan** [reply]
We plan to release at the end of Q3
Could you please tell how critical is this functionality for you?

**Reply by MhAhmadAli** [reply]
Well, it's pretty critical as it's part of a swap feature on the application.

I'll look for any alternative methods for the time being then.
